src/Chapter3.hs

```diff
@@ -479,6 +512,35 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city in total.
 -}
+data ChurchOrLibrary = Church | Library
+newtype House = House Int
+type Castle = Maybe String
+data Walls = Walls
+
+data City =  MkCity
+  { cityCastle          :: Castle,
+    cityHouses          :: [House],
+    cityCoL             :: ChurchOrLibrary,
+    cityWalls           :: Maybe Walls
+  }
+
+buildCastle :: City -> String -> City
+buildCastle city castleName =  city { cityCastle = Just castleName }
+
+buildHouse :: City -> House -> City
+buildHouse city@(MkCity _ houses _ _) newHouse@(House numPeople)
+  | numPeople > 0 && numPeople <= 4 = city { cityHouses = newHouse:houses }
+  | otherwise  = error "Number of people must be b/w 1-4."
+
+getHousePeople :: House -> Int
+getHousePeople (House x) = x;
+
+buildWalls :: City -> Walls -> City
+buildWalls (MkCity Nothing _ _ _) _ = error "No castle! Can't build walls."
```

Comment:
Usually, it is not a good idea to use `error`, because you can't easily catch such errors. In such situations you have multiple choices:

* Restrict input to disallow invalid states
* Don't change input on invalid arguments (i.e. return `City` without changes)
* Return a new custom data type and return it as a result, e.g.:
```haskell
data BuildWallsResult
    = NoCastle
    | NotEnoughPeople Int
    | NewCity City
```